libs/ml/include/ml/layers/fully_connected.hpp

```diff
@@ -82,26 +82,34 @@ class FullyConnected : public SubGraph<T>
     // start to set up the structure
     std::string input =
         this->template AddNode<fetch::ml::ops::PlaceHolder<TensorType>>(name + "_Input", {});
+    nodes_.at(input)->SetDefaultOutputShape({in_size_, 1});
 
     // for non time distributed layer, flatten the input
     std::string flat_input = input;
     if (!time_distributed_)
     {
       flat_input =
           this->template AddNode<fetch::ml::ops::Flatten<TensorType>>(name + "_Flatten", {input});
+      nodes_.at(flat_input)->SetDefaultOutputShape({out_size_, 1});
     }
 
     std::string weights =
         this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Weights", {});
+    nodes_.at(weights)->SetDefaultOutputShape({out_size_, in_size_});
     std::string weights_matmul = this->template AddNode<fetch::ml::ops::MatrixMultiply<TensorType>>(
         name + "_MatrixMultiply", {weights, flat_input});
+    nodes_.at(weights_matmul)->SetDefaultOutputShape({out_size_, 1});
     std::string bias =
         this->template AddNode<fetch::ml::ops::Weights<TensorType>>(name + "_Bias", {});
+    nodes_.at(bias)->SetDefaultOutputShape({out_size_, 1});
+
```

Comment:
i wonder if there's a way we can elegantly separate the layer construction work from the OutputShape setting work (without repeating code); having them interleaved like this seems suboptimal

Reply:
the idea behind this is: I know the shapes for all internal sub-nodes of a fully-connected layer, so there is no need to recursively traverse them and calculate shapes.

Reply:
Moved in a separate code block to prevent interleaving.